lib/identity_store.js

```diff
@@ -0,0 +1,129 @@
+        // console.log("headers2: "+http.defaults.headers.Authorization);
+        return this.http_api.get(this.base_url + endpoint, {params: {}})
+            .then(function(response) {
+                return response.data;
+            });
+    },
+
+    // Searches the Identity Store for all identities with the provided address.
+    // Returns the first identity object found
+    // Address should be an object {address_type: address}; eg.
+    // {'msisdn': '0821234444'}; {'email': 'me@example.com'}
+    search: function(address) {
+        var endpoint = 'identities/search/';
+
+        var address_type = Object.keys(address)[0];
+        var address_val = address[address_type];
+        var params = {};
+        var search_string = 'details__addresses__' + address_type;
+        params[search_string] = address_val;
+
+        return this.http_api.get(this.base_url + endpoint, {params: params})
+            .then(function(response) {
+                var identities_found = response.data.results;
+                // Return the first identity in the list of identities
+                return (identities_found.length > 0)
+                ? identities_found[0]
+                : null;
+            });
+    },
+
+    // Create a new identity
+    // Returns the identity object
+    create: function(address, options) {
+        var endpoint = 'identities/';
+
+        var payload = {
+            "details": {
+                "default_addr_type": null,
+                "addresses": {}
+            }
+        };
+        // compile base payload
+        if (address) {
+            var address_type = Object.keys(address);
+            var addresses = {};
+            addresses[address_type] = {};
+            addresses[address_type][address[address_type]] = {};
+            payload.details = {
+                "default_addr_type": "msisdn",
+                "addresses": addresses
+            };
+        }
+
+        if (options) {
+            if (options.communicate_through_id) {
+                payload.communicate_through = options.communicate_through_id;
+            }
+
+            // add operator_id if available
+            if (options.operator_id) {
+                payload.operator = options.operator_id;
+            }
+        }
+
+        return this.http_api.post(this.base_url + endpoint, {data: payload})
+            .then(function(response) {
+                return response.data;
+            });
+    },
+
+    // Gets a identity if it exists, otherwise creates a new one
+    get_or_create: function(address, options) {
+        var that = this;
+
+        return this.search(address)
+            .then(function(identity) {
+                if (identity !== null) {
+                    // If identity exists, return the id
+                    return identity;
+                } else {
+                    // If identity doesn't exist, create it
+                    return that.create(address, options)
```

Comment:
@imsickofmaps think it is due to something some might refers to as "binding loss"... something Javascript apparently suffers from.  See http://alistapart.com/article/getoutbindingsituations.  I tried quite a few ways of making this _get_or_create_ function work but I like this implementation most